platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/NodeId.java

```diff
@@ -16,148 +16,73 @@
 
 package com.swirlds.common.system;
 
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
 /**
  * A class that is used to uniquely identify a Swirlds Node
+ *
+ * @param id ID number unique within the network
  */
-public class NodeId implements EventCreationRule {
-    /** used to distinguish between a main node and a mirror node */
-    private boolean isMirror;
-    /** ID number unique within the network, unique set for main network and mirror network */
-    private long id;
+public record NodeId(long id) implements Comparable<NodeId> {
+
+    /** the map of all NodeId objects created, indexed by ID number */
+    private static final ConcurrentHashMap<Long, NodeId> nodeIds = new ConcurrentHashMap<>();
+
+    /** an invalid NodeId used in testing and boundary conditions. */
+    public static final NodeId INVALID_NODE_ID = NodeId.create(-1L);
+
+    /** the first NodeId object created */
+    public static final NodeId FIRST_NODE_ID = NodeId.create(0L);
 
     /**
      * Constructs a NodeId object
      *
-     * @param isMirror
-     * 		is it a mirror node or main node
-     * @param id
-     * 		the ID number
+     * @param id the ID number
      */
-    public NodeId(boolean isMirror, long id) {
-        this.isMirror = isMirror;
-        this.id = id;
+    public NodeId {
+        if (id < -1L) {
+            throw new IllegalArgumentException("id must be non-negative");
+        }
     }
 
     /**
      * Constructs a main network NodeId object
      *
-     * @param id
-     * 		the ID number
-     * @return the object created
-     */
-    public static NodeId createMain(long id) {
-        return new NodeId(false, id);
-    }
-
-    /**
-     * Constructs a mirror network NodeId object
-     *
-     * @param id
-     * 		the ID number
+     * @param id the ID number
      * @return the object created
      */
-    static NodeId createMirror(long id) {
-        return new NodeId(true, id);
+    public static NodeId create(long id) {
```

Comment:
I could turn it back into a class and make the constructor private.   The purpose of the static map is to eliminate the need for construction of multiple NodeIds wrapping the same long.   This reduces the need to grab new memory for the duplicate objects, and relieves the pressure on the garbage collector when the redundant objects lose scope.  Equality checks are short circuited by reference checks.

Reply:
Gotcha, I had misunderstood the purpose of the static map
 
I think I like a `class` with private constructor, to force usage of `create`. But since raw usage of the constructor wouldn't cause any problems (apart from performance), I think it's ok either way

Reply:
I have removed the caching of nodeId objects and construct a new object every time.